x/superfluid/keeper/stake_test.go

```diff
@@ -408,6 +408,152 @@ func (suite *KeeperTestSuite) TestSuperfluidUnbondLock() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestSuperfluidUndelegateAndUnbondLock() {
+	var lockAmount int64 = 1000000
+	testCases := []struct {
+		name            string
+		testInvalidLock bool
+		unlockAmount    sdk.Int
+		expectErr       bool
+		splitLockId     bool
+	}{
+		{
+			name:            "lock doesn't exist",
+			testInvalidLock: true,
+			unlockAmount:    sdk.NewInt(0),
+			expectErr:       true,
+			splitLockId:     false,
+		},
+		{
+			name:            "unlock amount = 0",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(0),
+			expectErr:       true,
+			splitLockId:     false,
+		},
+		{
+			name:            "unlock amount > locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount + 1),
+			expectErr:       true,
+			splitLockId:     false,
+		},
+		{
+			name:            "lock is not split if unlock amount = locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount),
+			expectErr:       false,
+			splitLockId:     false,
+		},
+		{
+			name:            "lock is split if unlock amount < locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount / 2),
+			expectErr:       false,
+			splitLockId:     true,
+		},
+	}
```

Comment:
Can we also add edge case testing for this method such as trying to UndelegateAndUnbondLock an existing unlocking lock || trying to UndelegateAndUnbondLock on a lock that still has bonding but is superfluid undelegating and so on?